Tools/pyco/pyco.py

```diff
@@ -453,44 +458,70 @@ class HasValue(Protocol):
 
 class Builder:
     def __init__(self):
-        self.codeobjs: list[CodeObject] = []
-        self.strings: list[String | Redirect] = []
-        self.blobs: list[BlobConstant] = []
-        self.constants: list[ComplexConstant] = []
+        SLOW_VERSION = False
+        if SLOW_VERSION:
+            self.codeobjs: list[CodeObject] = []
+            self.strings: list[String | Redirect] = []
+            self.blobs: list[BlobConstant] = []
+            self.constants: list[ComplexConstant] = []
+        else:
+            self.codeobjs: Mapping[CodeObject | Redirect, Tuple(int, int)] = OrderedDict()
+            self.strings: Mapping[String | Redirect, Tuple(int, int)] = OrderedDict()
+            self.blobs: Mapping[BlobConstant | Redirect, Tuple(int, int)] = OrderedDict()
+            self.constants: Mapping[ComplexConstant | Redirect, Tuple(int, int)] = OrderedDict()
         self.locked = False
-
         self.co_strings_start = -1
 
     def lock(self):
         self.locked = True
 
-    def add(self, where: list[T], thing: T) -> int:
+    def add(self, where: list[T|Redirect], thing: T|Redirect) -> int:
         # TODO: Avoid O(N**2) lookup behavior
         # Look for a match
-        for index, it in enumerate(where):
-            if type(it.value) is type(thing.value) and it.value == thing.value:
+        if isinstance(where, Mapping):
+            if thing in where:
+                index, _ = where[thing]
                 return index
+        else:
+            for index, it in enumerate(where):
+                if type(it.value) is type(thing.value) and it.value == thing.value:
+                    return index
         # Append a new one
         index = len(where)
         assert not self.locked
-        where.append(thing)
+        if isinstance(where, Mapping):
+            where[thing] = [index, index]
```

Comment:
Shouldn't that be a tuple?

Reply:
It was, but then I needed to re-assign the tuple in order to change the "last redirect", and that caused "modification during iteration" exceptions on the codeobjs array. So I made it array so that I can mutate the array without mutating the dict.